### ISSUE TITLE
Support CXSmiles as the output format in the cartridge.

### ISSUE THREAD

**Issue by greglandrum** [rdkit/rdkit#2795]
We can read and write it, but the output from pg_dump still uses normal SMILES.
We should change this to cxsmiles by default and add a GUC to allow it to be changed.

Related to #2791 

*Reading molecules from SMILES/CXSMILES*: This is easy since the RDKit's SMILES parser supports CXSMILES by default (if a SMILES string includes CXSMILES information it will be parsed).

*Explicitly generating SMILES/CXSMILES*: also easy, the functions `mol_to_smiles()` and `mol_to_cxsmiles()` are both defined for the cartridge

*Generating CXSMILES when a molecule is displayed as a string or when the database is dumped with `pg_dump`*: this is the primary point behind the ticket. The function that does this work for normal molecules (the `mol` type in postgresql) is called `mol_out()`. mol_out already generates CXSMILES. It's easiest to see this if a molecule has a conformation associated with it. Here's a quick example:

```
foodb=# select mol_to_cxsmiles(m) from mols_from_ctab where label='mol1';
                                 mol_to_cxsmiles                                 
---------------------------------------------------------------------------------
 c1ccccc1 |(1.5,0,;0.75,-1.299,;-0.75,-1.299,;-1.5,0,;-0.75,1.299,;0.75,1.299,)|
(1 row)

foodb=# select mol_out(m) from mols_from_ctab where label='mol1';
                                     mol_out                                     
---------------------------------------------------------------------------------
 c1ccccc1 |(1.5,0,;0.75,-1.299,;-0.75,-1.299,;-1.5,0,;-0.75,1.299,;0.75,1.299,)|
(1 row)
````
Notice that both commands produce the same output, `mol_to_smiles()` produces just normal SMILES:
```
foodb=# select mol_to_smiles(m) from mols_from_ctab where label='mol1';
 mol_to_smiles 
---------------
 c1ccccc1
(1 row)
```

Basically the key functionality required for to close this issue is already there, and I don't think it necessarily makes sense to add a GUC to changes the behavior

**Comment by manangoel99** [comment]
@greglandrum can I take this up?

**Comment by greglandrum** [comment]
@manangoel99: sure! 
The cartridge code is quite a bit different from the rest of the RDKit since it is primarily there to interface with PostgreSQL, so there's a bit of a learning curve.
If you want to give this a try, a good starting point would be to get a local cartridge build working and make sure the tests pass.

**Comment by greglandrum** [comment]
@manangoel99 : when you run cmake with the cartridge build enabled, there is a message displayed that tells you want you need to do to install it. Here's what I see:
```
After building, check
/scratch/RDKit_git/build_psql/Code/PgSQL/rdkit/pgsql_install.sh
for correctness of installation paths. If everything is OK, gain 
administrator privileges, stop the PostgreSQL service, run
/scratch/RDKit_git/build_psql/Code/PgSQL/rdkit/pgsql_install.sh
to install the PostgreSQL RDKit cartridge, then start again
the PostgreSQL service
```
once you've done that you can run the tests by running the pgsql_regress.sh script that is in the same directory as the pgsql_install.sh script

**Comment by manangoel99** [comment]
I built the cartridge with cmake but on running ```pgsql_reqression.sh``` , all the tests fail because when I looked at ```regression.diffs``` , the following error statement was present
```
ERROR:  syntax error at or near "@"
```

**Comment by greglandrum** [comment]
hmm, I've seen that one recently too. 
What OS are you running on and how did you install postgresql?

**Comment by manangoel99** [comment]
I'm using Ubuntu 19.04 and I installed postgresql using the steps given in https://www.digitalocean.com/community/tutorials/how-to-install-and-use-postgresql-on-ubuntu-18-04

**Comment by greglandrum** [comment]
ok, that's what I did too.
Something is broken here in the way the cartridge configure file is processed.
I will try and figure that out and get it fixed

**Comment by greglandrum** [comment]
@manangoel99 : it turns out that I had made a mess out of the cartridge config file and the CI builds didn't catch it.
#3073 has the fix and will hopefully be merged today

**Comment by manangoel99** [comment]
I'll wait for it to be merged and the continue working on this issue

**Comment by greglandrum** [comment]
it's merged

**Comment by manangoel99** [comment]
The local build is working and all the tests passed. How do I proceed further?

**Comment by greglandrum** [comment]
First some background so that this makes sense. I will eventually copy some of this up into the issue description to make it more obvious.

## Background on the cartridge.
Functionality is defined in three places:
1. `rdkit.sql.in`: This SQL template file is pre-processed when you run cmake and then copied to your postgresql extensions directory when you install the cartridge. It defines the types and functions that are available in the cartridge.
2. `rdkit_io.c`: A C file which contains the functions that the postgresql server actually calls when you invoke one of the RDKit SQL functions or use one of the RDKit extension types. The functions here have the same names as the functions defined in the SQL template file. These functions deal with the postgresql API in order to get the parameters passed to the function, call a C++ function defined in `adapter.cpp`, and then do whatever is required to return the results to postgresql. There are actually multiple `.c` files for the cartridge, but this is the one that's relevant to this ticket.
3. `adapter.cpp`: this contains the functions that are called by the C interface layer. The RDKit dependencies are all here.

## On to the ticket itself

Some background: 

*Reading molecules from SMILES/CXSMILES*: This is easy since the RDKit's SMILES parser supports CXSMILES by default (if a SMILES string includes CXSMILES information it will be parsed).

*Explicitly generating SMILES/CXSMILES*: also easy, the functions `mol_to_smiles()` and `mol_to_cxsmiles()` are both defined for the cartridge

*Generating CXSMILES when a molecule is displayed as a string or when the database is dumped with `pg_dump`*: this is the primary point behind the ticket. The function that does this work for normal molecules (the `mol` type in postgresql) is called `mol_out()`. mol_out already generates CXSMILES. It's easiest to see this if a molecule has a conformation associated with it. Here's a quick example:

```
foodb=# select mol_to_cxsmiles(m) from mols_from_ctab where label='mol1';
                                 mol_to_cxsmiles                                 
---------------------------------------------------------------------------------
 c1ccccc1 |(1.5,0,;0.75,-1.299,;-0.75,-1.299,;-1.5,0,;-0.75,1.299,;0.75,1.299,)|
(1 row)

foodb=# select mol_out(m) from mols_from_ctab where label='mol1';
                                     mol_out                                     
---------------------------------------------------------------------------------
 c1ccccc1 |(1.5,0,;0.75,-1.299,;-0.75,-1.299,;-1.5,0,;-0.75,1.299,;0.75,1.299,)|
(1 row)
````
Notice that both commands produce the same output, `mol_to_smiles()` produces just normal SMILES:
```
foodb=# select mol_to_smiles(m) from mols_from_ctab where label='mol1';
 mol_to_smiles 
---------------
 c1ccccc1
(1 row)
```

Wh

**Comment by greglandrum** [comment]
What should be done before closing the issue is to add an explicity test for this functionality. @manangoel99 : are you interested in doing that?

**Comment by manangoel99** [comment]
@greglandrum Sure. I'll be happy to do that

**Comment by greglandrum** [comment]
Ok, here's what you need to do.
The tests each consist of two files, one in `$RDBASE/Code/PgSQL/rdkit/sql` and one in `$RDBASE/Code/PgSQL/rdkit/expected`. In this case the tests should be added to `rdkit-91.sql` and the correponding output file.
The file in `sql` contains a series of SQL commands that should be executed and the corresponding file in `expected` contains the expected output.
Here you would want to be sure that when you construct a molecule from a CXSMILES with a conformation (some example above), that `mol_out()` produces a CXSMILES with information about the conformation.

Important note: you should edit the files in `$RDBASE/Code/PgSQL/rdkit/` because these are the ones in git. When you run cmake those files are copied into your build directory, so into `YOUR_BUILD_DIR/Code/PgSQL/rdkit/sql` and the corresponding `expected` directory
Don't forget to run cmake again (or copy the files manually) after you edit them and to look at the output of the tests to make sure your test was actually run.
